seastar/src/distributed.rs

```diff
@@ -0,0 +1,537 @@
+
+        fn stop(distr: &distributed) -> VoidFuture;
+    }
+}
+
+fn stop_caller<S: Service>(raw_service: *mut u8) -> VoidFuture {
+    VoidFuture::infallible_local(async move {
+        let stop_future = unsafe { (raw_service as *mut S).as_mut().unwrap().stop() };
+        Pin::from(stop_future).await;
+    })
+}
+
+const fn get_stop_caller<S: Service>() -> fn(*mut u8) -> VoidFuture {
+    stop_caller::<S>
+}
+
+use ffi::{distributed, VoidFuture};
+
+unsafe impl Send for distributed {}
+unsafe impl Sync for distributed {}
+
+/// A trait which a service inside `Distributed` must implement.
+///
+/// Because of Rust not yet supporting async trait methods,
+/// or trait methods that return an `impl` (`impl Future`) in this case,
+/// the returned future must be `Box`ed.
+///
+/// # Examples
+///
+/// ```rust
+/// use std::future::Future;
+/// use seastar::Service;
+///
+/// struct FooService;
+///
+/// impl Service for FooService {
+///     fn stop(&self) -> Box<dyn Future<Output = ()>> {
+///         Box::new(async { println!("Shutting down!") })
+///     }
+/// }
+/// ```
+pub trait Service {
+    /// The place to define what (possibly asynchronous) cleanup must be done for the service.
+    ///
+    /// If not implemented, defaults to a no-op.
+    fn stop(&self) -> Box<dyn Future<Output = ()>> {
+        Box::new(async {})
+    }
+}
+
+/// A service distributed amongst all shards of a Seastar app.
+///
+/// You can use this to, for example, load balance a local storage.
+pub struct Distributed<S: Service> {
+    _inner: SharedPtr<distributed>,
+    _ty: PhantomData<S>,
+}
+
+impl<S: Service> Distributed<S> {
+    pub fn start_inner<Func>(service_maker: Func, single: bool) -> impl Future<Output = Self>
+    where
+        Func: Fn() -> S + Sync,
+    {
+        let stop_caller = get_stop_caller::<S>();
+        let dropper = get_dropper_noarg::<S>();
+
+        let raw_service_maker = move || Box::into_raw(Box::new(service_maker())) as *mut u8;
+        let raw_service_maker_caller = get_fn_caller(&raw_service_maker);
+        let raw_service_maker_dropper = get_dropper(&raw_service_maker);
+        let boxed_raw_service_maker = Box::into_raw(Box::new(raw_service_maker)) as *mut u8;
+
+        let distr = ffi::new_distributed();
+        let fut = unsafe {
+            if single {
+                ffi::start_single(
+                    distr.as_ref().unwrap(),
+                    boxed_raw_service_maker,
+                    raw_service_maker_caller,
+                    stop_caller,
+                    dropper,
+                )
+            } else {
+                ffi::start(
+                    distr.as_ref().unwrap(),
+                    boxed_raw_service_maker,
+                    raw_service_maker_caller,
+                    stop_caller,
+                    dropper,
+                )
+            }
+        };
+
+        let maker_dropper = Dropper::new(boxed_raw_service_maker, raw_service_maker_dropper);
```

Comment:
Unfortunately, this doesn't solve the problem of "service_maker" being dropped too early. For example, the future returned by `start_inner` can be dropped immediately before "service_maker" has a chance to be called at all.

I think you should just transfer ownership of the service maker to the c++ side, and let `ffi::start`/`ffi::start_single` destroy it when it is appropriate.

Reply:
Can I send a clone of the maker to each shard and have each `rust_service` point to it and its dropper, and then free the maker's memory once it constructs itself?

Reply:
Sounds like it would work, but now you have to also teach C++ how to clone the maker. I don't think it's necessary to clone the maker, just make sure to destroy it after `distributed::start` finishes. So:

```c++
VoidFuture start(
    const distributed& distr,
    uint8_t* raw_service_maker,
    rust::Fn<uint8_t*(uint8_t*)> raw_service_maker_caller,
    /* raw_service_maker_dropper */,
    rust::Fn<VoidFuture(uint8_t*)> stop_caller,
    rust::Fn<void(uint8_t*)> dropper
) {
    // seastar::defer just runs the specified lambda in its destructor, so that it is cleaned up regardless of whether distributed::start throws an exception or not
    auto cleanup_maker = seastar::defer([&] {
        raw_service_maker_dropper(raw_service_maker);
    });
    co_await const_cast<distributed&>(distr).start(raw_service_maker, raw_service_maker_caller, /*raw_service_maker_dropper,*/ stop_caller, dropper);
}
```